### PR TITLE
M4156 Repository validation capabilities

### PR DESCRIPTION
**Pull request by dennishendriksen**

### CODE REVIEW

File: molgenis-data/src/main/java/org/molgenis/data/RepositoryCapability.java [hunk 1]
```diff
@@ -25,5 +25,20 @@
...
+
+	/**
+	 * Repository can validate unique values for attributes
+	 */
+	VALIDATE_UNIQUE_CONSTRAINT,
+
+	/**
+	 * Repository can validate if values are not-null for attributes
+	 */
+	VALIDATE_NOTNULL_CONTRAINT,
```
Comment (fdlk) [inline molgenis-data/src/main/java/org/molgenis/data/RepositoryCapability.java]: CONSTRAINT

**Comment by fdlk** [comment]
How are cross-repo references verified?

**Comment by fdlk** [comment]
The importer fails with a nullpointer exception when I import with mode add/update and it tries to figure out what row failed.

10:56:53.364 [pool-3-thread-2] INFO  org.molgenis.data.importer.ImportJob - Import failed.
java.lang.NullPointerException: null
	at org.molgenis.data.validation.ConstraintViolation.renumberRowIndex(ConstraintViolation.java:53) ~[classes/:na]
	at org.molgenis.data.validation.MolgenisValidationException.lambda$0(MolgenisValidationException.java:50) ~[classes/:na]
	at java.util.Collections$2.tryAdvance(Collections.java:4717) ~[na:1.8.0_71]
	at java.util.Collections$2.forEachRemaining(Collections.java:4725) ~[na:1.8.0_71]
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:580) ~[na:1.8.0_71]
	at org.molgenis.data.validation.MolgenisValidationException.renumberViolationRowIndices(MolgenisValidationException.java:50) ~[classes/:na]
	at org.molgenis.data.importer.ImportWriter.insertIntoRepo(ImportWriter.java:683) ~[classes/:na]
	at org.molgenis.data.importer.ImportWriter.update(ImportWriter.java:608) ~[classes/:na]
	at org.molgenis.data.importer.ImportWriter.importData(ImportWriter.java:228) ~[classes/:na]
	at org.molgenis.data.importer.ImportWriter.doImport(ImportWriter.java:122) ~[classes/:na]
	at org.molgenis.data.importer.ImportWriter$$FastClassBySpringCGLIB$$52fbfda4.invoke(<generated>) ~[spring-core-4.1.8.RELEASE.jar:na]
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204) ~[spring-core-4.1.8.RELEASE.jar:4.1.8.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:717) ~[spring-aop-4.1.8.RELEASE.jar:4.1.8.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157) ~[spring-aop-4.1.8.RELEASE.jar:4.1.8.RELEASE]
	at org.molgenis.security.core.runas.RunAsSystemProxy.invoke(RunAsSystemProxy.java:36) ~[classes/:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.1.8.RELEASE.jar:4.1.8.RELEASE]
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99) ~[spring-tx-4.1.8.RELEASE.jar:4.1.8.RELEASE]
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:281) ~[spring-tx-4.1.8.RELEASE.jar:4.1.8.RELEASE]
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96) ~[spring-tx-4.1.8.RELEASE.jar:4.1.8.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.1.8.RELEASE.jar:4.1.8.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:653) ~[spring-aop-4.1.8.RELEASE.jar:4.1.8.RELEASE]
	at org.molgenis.data.importer.ImportWriter$$EnhancerBySpringCGLIB$$dec8e50b.doImport(<generated>) ~[spring-core-4.1.8.RELEASE.jar:na]
	at org.molgenis.data.importer.EmxImportService.doImport(EmxImportService.java:83) ~[classes/:na]
	at org.molgenis.data.importer.EmxImportService.doImport(EmxImportService.java:70) ~[classes/:na]
	at org.molgenis.data.importer.ImportJob.run(ImportJob.java:51) ~[classes/:na]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_71]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_71]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_71]

**Comment by fdlk** [comment]
Error message not particularly clear yet:
```
; nested exception is org.springframework.jdbc.BadSqlGrammarException: PreparedStatementCallback; bad SQL grammar [SELECT COUNT(*) FROM "violationnotnull"]; nested exception is org.postgresql.util.PSQLException: ERROR: relation "violationnotnull" does not exist Position: 22
```

**Comment by fdlk** [comment]
Decorators try to remove capabilities from the list delivered by the underlying repository. But Postgres repo gives unmodifiable collection:

```
java.lang.UnsupportedOperationException: null
	at java.util.Collections$UnmodifiableCollection.remove(Collections.java:1058) ~[na:1.8.0_71]
	at org.molgenis.data.meta.MetaDataRepositoryDecorator.getCapabilities(MetaDataRepositoryDecorator.java:55) ~[classes/:na]
	at org.molgenis.data.EntityReferenceResolverDecorator.getCapabilities(EntityReferenceResolverDecorator.java:26) ~[classes/:na]
	at org.molgenis.data.ComputedEntityValuesDecorator.getCapabilities(ComputedEntityValuesDecorator.java:25) ~[classes/:na]
	at org.molgenis.ui.EntityListenerRepositoryDecorator.getCapabilities(EntityListenerRepositoryDecorator.java:54) ~[classes/:na]
	at org.molgenis.data.transaction.TransactionLogRepositoryDecorator.getCapabilities(TransactionLogRepositoryDecorator.java:52) ~[classes/:na]
	at org.molgenis.data.validation.RepositoryValidationDecorator.getCapabilities(RepositoryValidationDecorator.java:250) ~[classes/:na]
	at org.molgenis.data.AutoValueRepositoryDecorator.getCapabilities(AutoValueRepositoryDecorator.java:86) ~[classes/:na]
	at org.molgenis.data.RepositorySecurityDecorator.getCapabilities(RepositorySecurityDecorator.java:219) ~[classes/:na]
	at org.molgenis.data.importer.ImportWriter.reindex(ImportWriter.java:399) ~[classes/:na]
	at org.molgenis.data.importer.ImportWriter.rollbackSchemaChanges(ImportWriter.java:383) ~[classes/:na]
	at org.molgenis.data.importer.ImportWriter$$FastClassBySpringCGLIB$$52fbfda4.invoke(<generated>) ~[spring-core-4.1.8.RELEASE.jar:na]
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204) ~[spring-core-4.1.8.RELEASE.jar:4.1.8.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:717) ~[spring-aop-4.1.8.RELEASE.jar:4.1.8.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157) ~[spring-aop-4.1.8.RELEASE.jar:4.1.8.RELEASE]
	at org.molgenis.security.core.runas.RunAsSystemProxy.runAsSystem(RunAsSystemProxy.java:60) ~[classes/:na]
	at org.molgenis.security.core.runas.RunAsSystemProxy.invoke(RunAsSystemProxy.java:39) ~[classes/:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.1.8.RELEASE.jar:4.1.8.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:653) ~[spring-aop-4.1.8.RELEASE.jar:4.1.8.RELEASE]
	at org.molgenis.data.importer.ImportWriter$$EnhancerBySpringCGLIB$$dec8e50b.rollbackSchemaChanges(<generated>) ~[spring-core-4.1.8.RELEASE.jar:na]
	at org.molgenis.data.importer.EmxImportService.doImport(EmxImportService.java:91) [classes/:na]
	at org.molgenis.data.importer.EmxImportService.doImport(EmxImportService.java:70) [classes/:na]
	at org.molgenis.data.importer.ImportJob.run(ImportJob.java:51) [classes/:na]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_71]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_71]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_71]
```

**Comment by dennishendriksen** [comment]
Cross-repository collection reference validation

File: molgenis-data-validation/src/main/java/org/molgenis/data/validation/RepositoryValidationDecorator.java [hunk 2]
```diff
@@ -48,6 +48,8 @@
 import org.molgenis.util.HugeMap;
 import org.molgenis.util.HugeSet;
 
+import autovalue.shaded.com.google.common.common.base.Objects;
```
Comment (fdlk) [inline molgenis-data-validation/src/main/java/org/molgenis/data/validation/RepositoryValidationDecorator.java]: shaded copy of the class

File: molgenis-data-validation/src/main/java/org/molgenis/data/validation/RepositoryValidationDecorator.java [hunk 3]
```diff
@@ -362,43 +360,55 @@ private void initRequiredValueValidation(ValidationResource validationResource)
...
+		{
+			// validate cross-repository collection reference constraints. the decorated repository takes care of
+			// validating other reference constraints
+			String backend = getEntityMetaData().getBackend();
+			refAttrs = StreamSupport.stream(getEntityMetaData().getAtomicAttributes().spliterator(), false)
+					.filter(attr -> (attr.getDataType() instanceof XrefField || attr.getDataType() instanceof MrefField)
+							&& attr.getExpression() == null
+							&& !Objects.equal(attr.getRefEntity().getBackend(), backend))
+					.collect(Collectors.toList());
+		}
```
Comment (fdlk) [inline molgenis-data-validation/src/main/java/org/molgenis/data/validation/RepositoryValidationDecorator.java]: Nice!

**Pull request closed without merging**